rir/src/compiler/opt/constantfold.cpp

```diff
@@ -0,0 +1,100 @@
+#include "../pir/pir_impl.h"
+#include "../transform/replace.h"
+#include "../util/cfg.h"
+#include "../util/visitor.h"
+#include "R/Symbols.h"
+#include "pass_definitions.h"
+#include "../util/cfg.h"
+
+#include <unordered_set>
+
+namespace rir {
+namespace pir {
+
+#define FOLD(Instruction, Operation)                                           \
+    do {                                                                       \
+        if (auto instr = Instruction::Cast(i)) {                               \
+            if (auto lhs = LdConst::Cast(instr->arg<0>().val())) {             \
+                if (auto rhs = LdConst::Cast(instr->arg<1>().val())) {         \
+                    auto res = Rf_eval(Rf_lang3(Operation, lhs->c, rhs->c),    \
```

Comment:
I think this might be a protect bug, since `res` is neither protected nor inserted into the constant pool?

Edit: And are similar issues possible elsewhere in pir? This call to `eval` might collect anything that pir has created and not protected